### PR TITLE
Rework warning levels

### PR DESCRIPTION
**Pull request by RandomShaper**
Now you have: _extra_, _all_ (the default), _moderate_ and _no_.
The old _yes_ option is still supported for compatibility, but a warning about deprecation will be printed and _all_ will be assumed.

The different options will translate to the following:

- MSVC:
    - _extra_ -> `/Wall` (implies `/W4`)
    - _all_ -> `/W3` + disable non-essential (*)
    - _moderate_ -> `/W2` + disable non-essential (*)
    - _no_ -> `/w`

- GCC/Clang:
    - _extra_ -> `-Wall -Wextra`
    - _all_ -> `-Wall`
    - _moderate_ -> `-Wall -Wno-unused`
    - _no_ -> `-w`

(*) = Truncations, narrowing conversions...: `/wd4267 /wd4244 /wd4305 /wd4800`

This can probably be more fine-tuned, but in the meantime it improves the current state of things substantially, with that _TODO_ promising more options than on/off.

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -272,16 +272,26 @@ if selected_platform in platform_list:
...
-    # TODO: Add support to specify different levels of warning, e.g. only critical/significant, instead of on/off
-    if (env["warnings"] == "yes"):
-        if (os.name == "nt" and os.getenv("VSINSTALLDIR")): # MSVC, needs to stand out of course
-	   pass# env.Append(CCFLAGS=['/W2'])
-        else: # Rest of the world
-            env.Append(CCFLAGS=['-Wall'])
-    else:
-        if (os.name == "nt" and os.getenv("VSINSTALLDIR")): # MSVC
+    if (env["warnings"] == 'yes'):
+        prtint("WARNING: warnings=yes is deprecated; assuming warnings=all")
```
Comment (akien-mga) [inline SConstruct]: Typo? :) `prtint`.

File: SConstruct [hunk 2]
```diff
@@ -272,16 +272,26 @@ if selected_platform in platform_list:
...
+
+    if (os.name == "nt" and os.getenv("VSINSTALLDIR")): # MSVC, needs to stand out of course
+        if (env["warnings"] == 'extra'):
+            env.Append(CCFLAGS=['/Wall'])
+        elif (env["warnings"] == 'all' or env["warnings"] == 'yes'):
+            env.Append(CCFLAGS=['/W4'])
```
Comment (akien-mga) [inline SConstruct]: I initially put `/W4` by default, and MSVC users freaked out so @reduz disabled *all* warnings. I guess MSVC needs a lower warning level.

File: SConstruct (hunk 1 above)
Comment (RandomShaper) [inline SConstruct]: Good catch!

File: SConstruct (hunk 2 above)
Comment (RandomShaper) [inline SConstruct]: Now the default would be `/W3` which you can lower to `/W2` with _moderate_. What about that?

**Review by akien-mga — APPROVED** [review]

**Comment by reduz** [comment]
deja vu.. I remember the screen being flooded with pointless warnings in
MSVC last time this changed :P

**Comment by RandomShaper** [comment]
I've added some more relief about this kind

> core\math\face3.h(240): warning C4305: 'initializing': truncation from 'double' to 'real_t'

and this kind

> core\math\math_2d.h(418): warning C4244: 'argument': conversion from 'const int' to 'real_t', possible loss of data

One should keep an eye on this things from time to time. But they are probably not relevant for the style/needs of Godot so I'm disabling them by default.

**Comment by reduz** [comment]
i don' t recally any bug ever as a result of this, or signed-unsigned
comparisons.. so having to do all casts manually for bugs that never happen
is a hassle :\

**Comment by RandomShaper** [comment]
I hope now you find the default warning settings are sensible. About truncations, it stills shows warnings when they involve constants, but very few.

**Comment by RandomShaper** [comment]
I've just built with MSVC with the last settings I've come up with and it's overall clean.

The biggest offender is `warning C4800: '...': forcing value to bool 'true' or 'false' (performance warning)` which is interesting. I may even fix them myself.

**Comment by RandomShaper** [comment]
Guess what? After reading Microsoft's explanation, I don't think it's relevant for us neither there's anything to fix about it, so disabling as well.

Don't worry! I think this is the last iteration of this PR. :)

**Comment by akien-mga** [comment]
Ah, gotta love GCC and Clang for meaningful warning levels :P

**Pull request merged**